lib/upload_to_s3.js

```diff
@@ -1,45 +1,84 @@
 import https from 'https';
 import url from 'url';
+import fs from 'fs';
+import temporary from 'temporary';
+import promiseRetry from 'promise-retry';
+import extend from 'extend';
 
-export default async function uploadToS3 (task, sourceStream, artifactName, expiration, httpsHeaders) {
-  var queue = task.runtime.queue;
-
-  var artifact = await queue.createArtifact(
-    task.status.taskId,
-    task.runId,
-    artifactName,
-    {
-      // Why s3? It's currently cheaper to store data in s3 this could easily
-      // be used with azure simply by changing s3 -> azure.
-      storageType: 's3',
-      expires: new Date(Math.min(expiration, new Date(task.task.expires))),
-      contentType: httpsHeaders['content-type']
-    }
-  );
-
-  let parsedUrl = url.parse(artifact.putUrl);
-  let options = {
-    hostname: parsedUrl.hostname,
-    path: parsedUrl.path,
-    method: 'PUT',
-    headers: httpsHeaders
-  };
-
-  let req = https.request(options);
-  sourceStream.pipe(req);
-
-  let response = await new Promise((accept, reject) => {
-    req.on('response', accept);
-    req.on('error', reject);
-    req.setTimeout(5 * 60 * 1000, reject);
+export default async function uploadToS3 (
+  task,
+  sourceStream,
+  artifactName,
+  expiration,
+  httpsHeaders,
+  putUrl,
+  additionalOptions)
+{
+  let queue = task.runtime.queue;
+
+  let tmp = new temporary.File();
```

Comment:
Things are actually more complicated. Temporary file names suffer from the [TOCTTOU](https://en.wikipedia.org/wiki/Time_of_check_to_time_of_use). The only reliable way to generate temporary files is by using the OS corresponding system call, like [tmpfile](http://linux.die.net/man/3/tmpfile) or [GetTempFileName](https://msdn.microsoft.com/en-us/library/aa364991(v=vs.85).aspx). I look at several nodejs packages but no one implemented a reliable temp file function. In summary, anyway we are theoretically screwed.